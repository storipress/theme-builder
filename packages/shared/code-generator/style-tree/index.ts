import produce, { freeze, original } from 'immer'
import { merge, pickBy } from 'lodash'
import invariant from 'tiny-invariant'

export interface Breakpoint {
  xs: unknown
  md?: unknown
  lg?: unknown
}

export type BREAKPOINTS = keyof Breakpoint

// it must be cloneable for pass across iframe
export const STYLE_TREE_FRAGMENT = '@@jraft/STYLE_TREE_FRAGMENT'

export const BREAKPOINTS = ['xs', 'md', 'lg'] as (keyof Breakpoint)[]

export interface StyleTreeMeta {
  dirty: Record<string, BREAKPOINTS>
}

export interface StyleTree {
  name: string | typeof STYLE_TREE_FRAGMENT
  styles: Record<string, Breakpoint>
  children: Record<string, StyleTree>
  meta?: StyleTreeMeta
}

export function createStyleTree(name: string = STYLE_TREE_FRAGMENT): StyleTree {
  return {
    name,
    styles: {},
    children: {},
  }
}

export function filterChildren(tree: StyleTree, predictor: (key: string) => boolean): StyleTree {
  return produce(tree, (draft) => {
    draft.children = pickBy(draft.children, (_, key) => predictor(key))
  })
}

export function assertStyleTree(tree: StyleTree, root?: string): StyleTree {
  if (!tree || root === '' || (root && tree.name !== root)) {
    return freeze(createStyleTree(root || STYLE_TREE_FRAGMENT))
  }
  return produce(tree, (draft) => {
    ensureStyleTreeFormat(draft)
  })
}

function ensureStyleTreeFormat(tree: StyleTree) {
  if (typeof tree.styles !== 'object' || Array.isArray(tree.styles)) {
    tree.styles = {}
  }

  if (typeof tree.children !== 'object' || Array.isArray(tree.children)) {
    tree.children = {}
  }

  for (const subtree of Object.values(tree.children)) {
    ensureStyleTreeFormat(subtree)
  }
}

export function ensureNode(root: StyleTree | undefined, path: readonly string[]): StyleTree {
  if (path.length === 0) {
    invariant(root)
    return root
  }

  if (!root) {
    root = {
      name: path[0],
      styles: {},
      children: {},
    }
  }

  if (root.name !== STYLE_TREE_FRAGMENT) {
    invariant(path[0] === root.name, 'unmatch path name')
    const [, ...rest] = path
    if (rest.length > 0) {
      root.children[rest[0]] = ensureNode(root.children[rest[0]], rest)
    }
    return root
  }
  root.children[path[0]] = ensureNode(root.children[path[0]], path)
  return root
}

export function ensureStyle(root: StyleTree, path: readonly string[]): Record<string, Breakpoint> {
  ensureNode(root, path)
  const node = getNode(root, path)
  return node.styles
}

export function getStyleWithDefault(root: StyleTree, p: readonly string[]): Record<string, Breakpoint> {
  const node = getOptionalNode(root, p) ?? createStyleTree()
  return node.styles
}

export function getNode(root: StyleTree, p: readonly string[]): StyleTree {
  const node = getOptionalNode(root, p)
  invariant(node, 'node not found')
  return node
}

export function getOptionalNode(root: StyleTree, p: readonly string[]): StyleTree | undefined {
  let node = root
  const path = [...p]

  if (path.length === 0) {
    return root
  }

  if (!root) {
    return
  }

  if (root.name === STYLE_TREE_FRAGMENT) {
    node = node.children[path[0]]
  }

  if (!node) {
    return
  }

  while (path.length > 0) {
    invariant(path[0] === node.name, 'unmatch path name')
    path.shift()
    if (path.length === 0) {
      break
    }
    node = node.children[path[0]]
    if (!node) {
      return node
    }
  }
  return node
}

export function mergeTree(base?: StyleTree, tree?: StyleTree): StyleTree {
  invariant(base || tree, 'both empty')
  if (!tree) {
    return base as StyleTree
  }
  if (!base) {
    return tree
  }
  const node = createStyleTree(base.name)
  node.styles = merge({}, base.styles, tree.styles)
  const children = [...new Set([...Object.keys(base.children), ...Object.keys(tree.children)])]
  for (const key of children) {
    node.children[key] = mergeTree(base.children[key], tree.children[key])
  }
  return node
}

export function detachTree(tree: StyleTree, path: readonly string[]): { left: StyleTree; right: StyleTree } {
  if (path.length === 0) {
    return { left: tree, right: createStyleTree() }
  }
  let right = createStyleTree()
  const p = path.slice(0, -1)
  const last = path[path.length - 1]
  const left = produce(tree, (draft) => {
    const node = getOptionalNode(draft, p)
    if (!node || !node.children[last]) {
      return
    }
    right = original(node.children[last]) as StyleTree
    Reflect.deleteProperty(node.children, last)
  })
  return { left, right }
}

export function scopeTree(scope: string, tree: StyleTree): StyleTree {
  return {
    name: scope,
    styles: {},
    children: {
      [tree.name]: tree,
    },
  }
}

export const renderHelpers = {
  STYLE_TREE_FRAGMENT,

  useDeepSelector: false,

  ensureSelector(maybeSelector: string): string {
    if (this.hasElementSelector(maybeSelector)) {
      return this.renderElementSelector(maybeSelector)
    }
    return `.${maybeSelector}`
  },

  hasElementSelector(selector: string): boolean {
    return selector.startsWith('&')
  },

  renderElementSelector(selector: string): string {
    return this.useDeepSelector ? selector.replace('&', '&::v-deep') : selector
  },
}

export function insertToTree(
  tree: StyleTree,
  path: readonly string[],
  data: Record<string, unknown>,
  breakpoint: keyof Breakpoint = 'xs',
  noOverride: boolean = false
): StyleTree {
  const style = ensureStyle(tree, path)
  const normalized = normalizeStyles(data, breakpoint)
  insertStyles(style, normalized, noOverride)
  return tree
}

export function insertToTreeWithCascadeDown(
  tree: StyleTree,
  path: readonly string[],
  raw: Record<string, unknown>,
  breakpoint: keyof Breakpoint = 'xs',
  noOverride: boolean = false
): StyleTree {
  ensureNode(tree, path)
  const node = getNode(tree, path)
  const { styles, meta } = node
  const { styles: normalized, meta: newMeta } = cascadeDownStyles({ raw, meta, breakpoint })
  insertStyles(styles, normalized, noOverride)
  node.meta = newMeta
  return tree
}

export function insertStyles(
  styles: Record<string, Breakpoint>,
  newStyles: Record<string, Breakpoint>,
  noOverride: boolean = false
): Record<string, Breakpoint> {
  for (const [key, value] of Object.entries(newStyles)) {
    for (const breakpoint of BREAKPOINTS) {
      // null means inherit/unset
      if (styles[key]?.[breakpoint] === null || value[breakpoint] === undefined) {
        continue
      }
      styles[key] ??= {} as any

      if (styles[key][breakpoint] != null && value[breakpoint] === null) {
        continue
      }

      if (styles[key][breakpoint] != null && noOverride) {
        continue
      }

      styles[key][breakpoint] = value[breakpoint]
    }
  }
  return styles
}

const CASCADE_DOWN: Record<BREAKPOINTS, BREAKPOINTS[]> = {
  lg: ['lg', 'md', 'xs'],
  md: ['md', 'xs'],
  xs: ['xs'],
}
interface CascadeDownInput {
  raw: Record<string, unknown>
  meta?: StyleTreeMeta
  breakpoint?: BREAKPOINTS
}
interface CascadeDownReturn {
  meta: StyleTreeMeta
  styles: Record<string, Breakpoint>
}

export function cascadeDownStyles({ raw, meta, breakpoint = 'xs' }: CascadeDownInput): CascadeDownReturn {
  const styles = normalizeStyles(raw, breakpoint)
  const { dirty = {} } = meta ?? {}
  for (const [key, style] of Object.entries(styles)) {
    const d = dirty[key] ?? 'lg'
    if (d !== breakpoint && CASCADE_DOWN[breakpoint].includes(d)) {
      continue
    }
    for (const cascade of CASCADE_DOWN[breakpoint]) {
      style[cascade] = style[breakpoint]
    }
    dirty[key] = breakpoint
  }

  return { meta: { dirty }, styles }
}

export function normalizeStyles(
  original: Record<string, unknown>,
  breakpoint: keyof Breakpoint = 'xs'
): Record<string, Breakpoint> {
  const res: Record<string, Breakpoint> = {}

  for (const key of Object.keys(original)) {
    res[key] = normalizeStyle(original[key], breakpoint)
  }
  return res
}

export function normalizeStyle(data: unknown, breakpoint: keyof Breakpoint = 'xs'): Breakpoint {
  if (data && typeof data === 'object') {
    return data as Breakpoint
  }
  return { [breakpoint]: data } as Partial<Breakpoint> as Breakpoint
}

export function pluckBreakpoint(
  styles: Record<string, Breakpoint>,
  breakpoint: keyof Breakpoint
): Record<string, unknown> {
  const res: Record<string, unknown> = {}
  for (const key of Object.keys(styles)) {
    res[key] = styles[key][breakpoint] ?? (breakpoint === 'lg' ? styles[key].md : undefined) ?? styles[key].xs
  }
  return res
}
