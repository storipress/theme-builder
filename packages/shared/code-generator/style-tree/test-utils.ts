import type { Breakpoint, StyleTree } from '.'
import produce from 'immer'

import invariant from 'tiny-invariant'
import { getNode, STYLE_TREE_FRAGMENT } from '.'

export interface Options {
  getStore: () => Record<string, StyleTree>
  key?: string
}

export interface UpdateOptions {
  path: readonly string[]
  property: string
  value: unknown
}

interface NodeInfo {
  node: StyleTree
  name: string
  path: string[]
  apply: (styles: Record<string, Breakpoint>) => Promise<void>
}

let lastStore: unknown

export interface TestUtils {
  resolveSelector: (path: readonly string[]) => string
  nextTick: () => Promise<void>
  update: (options: UpdateOptions) => Promise<() => void>
  traverse: (cb: (info: NodeInfo) => Promise<void>) => Promise<void>
}

export function resolveSelector(path: readonly string[]): string {
  invariant(path.length > 0, 'expect a least one selector')
  const name = path[path.length - 1]
  if (name.startsWith('&')) {
    invariant(path.length > 1, 'expect has parent')
    return `${resolveSelector(path.slice(0, -1))} ${name.replace('& ', '')}`
  }
  return `.${name}`
}

function assertStore(store: unknown) {
  if (!lastStore) {
    store = lastStore
    return
  }
  invariant(store === lastStore, 'found multiple instance of store')
}

export function createTestUtils({ getStore, key = 'styles' }: Options): TestUtils {
  return {
    resolveSelector,
    nextTick,

    async update({ path, property, value }: UpdateOptions) {
      const store = getStore()
      let old: Breakpoint
      store[key] = produce(store[key], (root) => {
        const node = getNode(root, path)
        old = node.styles[property]
        node.styles[property] = typeof value !== 'object' ? { xs: value } : (value as Breakpoint)
      })

      return () => {
        store[key] = produce(store[key], (root) => {
          const node = getNode(root, path)
          node.styles[property] = old
        })
      }
    },

    async traverse(handler) {
      const store = getStore()
      assertStore(store)
      const root = store[key]
      const queue = [[root, [] as string[]]]
      const shouldSkip = new Set()
      while (queue.length > 0) {
        const [node, path] = queue.shift() as [StyleTree, string[]]
        if (node === root && node.name === STYLE_TREE_FRAGMENT) {
          for (const child of Object.values(node.children)) {
            shouldSkip.add(child)
            queue.push([child, [...path, child.name]])
          }
          continue
        }

        for (const child of Object.values(node.children)) {
          queue.push([child, [...path, child.name]])
        }

        if (shouldSkip.has(node)) {
          shouldSkip.delete(node)
          continue
        }

        if (node.name === STYLE_TREE_FRAGMENT) {
          continue
        }

        try {
          await handler({
            node,
            name: node.name,
            path,
            apply: async (styles: Record<string, Breakpoint>) => {
              store[key] = produce(root, (draft) => {
                const draftNode = getNode(draft, path)
                draftNode.styles = styles
              })
              await nextTick()
            },
          })
        } catch (error) {
          console.error(error)
        }
      }
      store[key] = root
    },
  }
}

function nextTick(): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, 0)
  })
}
