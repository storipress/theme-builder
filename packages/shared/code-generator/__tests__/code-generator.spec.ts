import { generate } from '..'
import { createStyleTree } from '../style-tree'

describe('code-generator', () => {
  describe('generate', () => {
    it('return css', () => {
      const tree = createStyleTree('foo')
      tree.styles = {
        fontSize: { xs: '12px', lg: '14px', md: '16px' },
      }
      expect(generate(tree)).toMatchSnapshot()

      const tree2 = createStyleTree('foo')
      tree2.children = {
        '& p': {
          name: '& p',
          styles: {
            fontSize: { xs: '12', lg: '14', md: '16' },
          },
          children: {},
        },
      }
      expect(generate(tree2)).toMatchSnapshot()
    })

    it('could use null style', () => {
      const tree = createStyleTree('foo')
      tree.styles = {
        fontSize: { xs: '12px', lg: '14px', md: null },
      }
      expect(generate(tree)).toMatchSnapshot()
    })

    it('could use deep selector', () => {
      const tree = createStyleTree('foo')
      tree.children = {
        '& p': {
          name: '& p',
          styles: {
            fontSize: { xs: '12', lg: '14', md: '16' },
          },
          children: {},
        },
      }
      const css = generate(tree, true)
      expect(css).toContain('v-deep')
      expect(css).toMatchSnapshot()
    })
  })
})
