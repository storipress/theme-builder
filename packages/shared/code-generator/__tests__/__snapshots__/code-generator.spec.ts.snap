// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`code-generator > generate > could use deep selector 1`] = `".foo::v-deep p{font-size:12px;}@media (min-width: 768px) {.foo::v-deep p{font-size:16px}}@media (min-width: 1070px) {.foo::v-deep p{font-size:14px}}"`;

exports[`code-generator > generate > could use null style 1`] = `
".foo{font-size:12pxpx;}
@media (min-width: 1070px) {.foo{font-size:14pxpx}}"
`;

exports[`code-generator > generate > return css 1`] = `".foo{font-size:12pxpx;}@media (min-width: 768px) {.foo{font-size:16pxpx}}@media (min-width: 1070px) {.foo{font-size:14pxpx}}"`;

exports[`code-generator > generate > return css 2`] = `".foo p{font-size:12px;}@media (min-width: 768px) {.foo p{font-size:16px}}@media (min-width: 1070px) {.foo p{font-size:14px}}"`;
